doc/decisions/creating_keys.md

```diff
@@ -0,0 +1,145 @@
+# Creating Keys
+
+## Problem
+
+The API that `libelektra-core` defines for creating Keys must be possible to implement in other languages without complicated workarounds.
+In particular, another implementation should not need to resort to adding a small amount of C to "bridge the gap".
+
+## Constraints
+
+- There should be a way to create a `Key *` with name, value and metadata in a single expression.
+
+  This is needed to allow creating a full `KeySet *` with a single expression, which in turn is very useful e.g. when initializing default configurations.
+
+- Using the escaped name in the constructor function should be avoided, since the unescaping process can be expensive.
+- Ideally, creating a `Key *` without metadata should use at most 3 `malloc`s and 2 `memcpy` to:
+
+  1. `malloc` to allocate the `struct _Key`.
```

Comment:
> There may of course be optimizations for smaller values.

Which is a common case.

> so that even creating multiple keys only calls malloc once.

I don't think this is relevant. We have mmap for that.

Reply:
> > There may of course be optimizations for smaller values.
> 
> Which is a common case.

I agree that optimizations are useful and that small key names/values are a common case. However, this is entirely irrelevant here. The constraint is supposed to describe the general case that holds **for all** cases. In other words it is an upper bound. (*)



> > so that even creating multiple keys only calls malloc once.
> 
> I don't think this is relevant. We have mmap for that.

Yes, `mmapstorage` does something similar, but arena allocation is more general. `mmapstorage` only works when you load from a config file. But arena allocation could be used for keys created at runtime too. We would of course need benchmarks to prove this, but I in my experience there is a lot of code that calls `keyNew`/`ksDel` and then calls `keyDel`/`ksDel` shortly afterwards. All of these currently are `malloc`/`free` calls. With arena allocation we could minimize that. It is of course a very complicated topic and not worth doing without intensive benchmarks, but it was just hypothetical.

---

(*) It's not a hard bound, however. The bound may be violated, _if_ there is a good technical reason (e.g. adding another malloc/memcpy makes the code a lot simpler and benchmarks show there is no significant impact). That is what I meant with "ideally". I'll reformulate this part to make the intentions more clear.

Reply:
> this is entirely irrelevant here

Based on which criteria is this irrelevant? I thought we are discussing to remove the arguments of keyNew? The reason of the current arguments of keyNew is that keyNew could allocate name+value in one malloc (I unfortunately did not write decisions about this but luckily we write them now).

> there is a lot of code that calls keyNew/ksNew [fixed typo] and then calls keyDel/ksDel

Yes, I agree this is a problem, especially because at the moment the whole unescaping needs to be done for every keyNew. I looked at a few places and one problems seems to be that only ksLookup provides an API where a key name can be directly given. This could be solved by a KeyName struct (which I do not like) but it could also be solved by making keyNew cheaper. In previous benchmarks basically only (un)escaping was expensive not the malloc.

Ideally, only one keyNew for the parentKey is needed and everything else is mmaped anyway.

Reply:
> Based on which criteria is this irrelevant?

The fact that _some_ cases of "creating a key with name and value" can be optimized to use less than 3 `malloc` or less than 2 `memcpy` is irrelevant here, because the decision is about the general case. The general case (arbitrary name, arbitrary value) cannot (without extra precautions) use less than what this decision already states.

> keyNew could allocate name+value in one malloc

In theory, you could even allocate struct+name+value in one malloc. However, this may require changes elsewhere, because `realloc`/`free` will no longer work, if name/value points into the middle of a malloc-buffer. In particular, you need to use some flags (maybe the mmap ones) to indicate which pointers are from `malloc` and need to be `free`d.

> his could be solved by a KeyName struct (which I do not like) but it could also be solved by making keyNew cheaper. In previous benchmarks basically only (un)escaping was expensive not the malloc.

I agree, making `keyNew` cheaper is the main goal.

---

To sum up:

I will rewrite this part of the decision to make it more clear that

1. `keyNew` should be as cheap as possible, because it may be called very often.
2. The `keyNew` API should be designed such that `malloc` and `memcpy` can be used as little as possible.